packages/next-auth/src/lib/index.ts

```diff
@@ -0,0 +1,238 @@
+   * Callbacks **allow you to implement access controls without a database** or to **integrate with external databases or APIs**.
+   */
+  callbacks?: AuthConfig["callbacks"] & {
+    /**
+     * Invoked when a user needs authorization, using [Middleware](https://nextjs.org/docs/advanced-features/middleware).
+     *
+     * You can override this behavior by returning a {@link NextResponse}.
+     *
+     * @example
+     * ```ts title="app/auth.ts"
+     * ...
+     * async authorized({ request, auth }) {
+     *   const url = request.nextUrl
+     *
+     *   if(request.method === "POST") {
+     *     const { authToken } = (await request.json()) ?? {}
+     *     // If the request has a valid auth token, it is authorized
+     *     const valid = await validateAuthToken(authToken)
+     *     if(valid) return true
+     *     return NextResponse.json("Invalid auth token", { status: 401 })
+     *   }
+     *
+     *   // Logged in users are authenticated, otherwise redirect to login page
+     *   return !!auth.user
+     * }
+     * ...
+     * ```
+     *
+     * :::warning
+     * If you are returning a redirect response, make sure that the page you are redirecting to is not protected by this callback,
+     * otherwise you could end up in an infinite redirect loop.
+     * :::
+     */
+    authorized?: (params: {
+      /** The request to be authorized. */
+      request: NextRequest
+      /** The authenticated user or token, if any. */
+      auth: AuthSession | null
+    }) => Awaitable<boolean | NextResponse | undefined>
+  }
+}
+
+/** Server-side method to read the session. */
+async function getSession(headers: Headers, config: NextAuthConfig) {
+  const origin = detectOrigin(headers)
+  const request = new Request(`${origin}/api/auth/session`, {
+    headers: { cookie: headers.get("cookie") ?? "" },
+  })
+  config.useSecureCookies ??= origin.protocol === "https:"
+
+  return Auth(request, {
+    ...config,
+    callbacks: {
+      ...config.callbacks,
+      // Since we are server-side, we don't need to filter out the session data
+      // See https://nextjs.authjs.dev/v5#authenticating-server-side
+      session({ session, user, token }) {
+        return { ...session, user: user ?? token }
+      },
+    },
+  })
+}
+
+export interface NextAuthRequest extends NextRequest {
+  auth: AuthSession | null
+}
+
+export type NextAuthMiddleware = (
+  request: NextAuthRequest,
+  event: NextFetchEvent
+) => ReturnType<NextMiddleware>
+
+export type WithAuthArgs =
+  | [NextAuthRequest, any]
+  | [NextAuthMiddleware]
+  | [AppRouteHandlerFn]
+  | [NextApiRequest, NextApiResponse]
+  | [GetServerSidePropsContext]
+  | []
+
+function isReqWrapper(arg: any): arg is NextAuthMiddleware | AppRouteHandlerFn {
+  return typeof arg === "function"
+}
+
+export function initAuth(config: NextAuthConfig) {
+  return (...args: WithAuthArgs) => {
+    if (!args.length) return getSession(headers(), config).then((r) => r.json())
```

Comment:
Hey! Was just wondering if this return should be wrapped in a `cache()` so that calling `auth()` from multiple server components would only do the calculation once per request? If not, recommending that as an option in the docs may be a good idea.

Reply:
something as simple as
```ts
const getRSCSession = cache((config: NextAuthConfig) => getSession(headers(), config).then((r) => r.json()))

// ...

if (!args.length) return getRSCSession(config)
```